sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,448 @@
+                x_squared_norms=x_squared_norms, n_threads=self._n_threads
+            )
+
+            if best_inertia is None or inertia < best_inertia:
+                best_labels = labels
+                best_centers = centers
+                best_inertia = inertia
+                best_n_iter = n_iter_
+
+        if not sp.issparse(X):
+            if not self.copy_x:
+                X += X_mean
+            best_centers += X_mean
+
+        distinct_clusters = len(set(best_labels))
+        if distinct_clusters != 2:
+            warnings.warn(
+                "Number of distinct clusters ({}) found smaller than "
+                "n_clusters ({}). Possibly due to duplicate points "
+                "in X.".format(distinct_clusters, 2),
+                ConvergenceWarning, stacklevel=2)
+
+        return best_centers, best_labels, best_inertia, best_n_iter
+
+    def fit(self, X, y=None, sample_weight=None):
+        """Compute bisecting k-means clustering.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+
+            Training instances to cluster.
+
+            .. note:: The data will be converted to C ordering,
+                which will cause a memory copy
+                if the given data is not C-contiguous.
+
+        y : Ignored
+            Not used, present here for API consistency by convention.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            The weights for each observation in X. If None, all observations
+            are assigned equal weight.
+
+        Returns
+        -------
+        self
+            Fitted estimator.
+        """
+        X = self._validate_data(X, accept_sparse='csr',
+                                dtype=[np.float64, np.float32],
+                                order='C', copy=self.copy_x,
+                                accept_large_sparse=False)
+
+        self._check_params(X)
+        random_state = check_random_state(self.random_state)
+        sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
+        self._n_threads = _openmp_effective_n_threads()
+
+        if self._algorithm == "full":
+            self._kmeans_single = _kmeans_single_lloyd
+            self._check_mkl_vcomp(X, X.shape[0])
+        else:
+            self._kmeans_single = _kmeans_single_elkan
+
+        data_left = X
+        weights_left = sample_weight
+
+        centroids = []
+
+        if self.verbose:
+            print("Running Bisecting K-Means with parameters:")
+            print("-> number of clusters: {}".format(self.n_clusters))
+            print("-> number of centroid initializations: {}"
+                  .format(self.n_init))
+            print("-> relative tolerance: {:.4e} \n".format(self.tol))
+
+        for n_iter in range(self.n_clusters):
+
+            # Perform Bisection
+            centers, labels, _,  _ = self._bisect(data_left, y, weights_left,
```

Comment:
You should not make `_bisect` return four variables if you only use two.

Reply:
I left that in case it might be potentially needed in future for debug or used with `verbose`. But if you say so - I'll remove it